plugins/disk/nvme

```diff
@@ -0,0 +1,244 @@
+diskstats.  If they are much higher, look into if the write
+amplification can be due to suboptimal I/O request sizes.
+
+=head2 nvme_writecycles
+
+This graphs is intended to give an indication of how much life there
+is left in your NVMe.  It calculates the number of bytes written
+during each device's lifetime against the capacity of the device,
+thereby getting an average number of write cycle each cell has
+experienced.
+
+A prosumer NVMe will handle a few thousand writes to each cell before
+the error rate gets out of hand.
+
+=head1 MAGIC MARKERS
+
+  #%# family=auto
+  #%# capabilities=autoconf
+
+=head1 BUGS
+
+None known.
+
+=head1 VERSION
+
+  1.0
+
+=head1 AUTHOR
+
+Kjetil Torgrim Homme <kjetil.homme@redpill-linpro.com>
+
+=head1 LICENSE
+
+GPLv2
+
+=cut
+
+use strict;
+use Munin::Plugin;
+
+# Check that multigraph is supported
+need_multigraph();
+
+sub run_nvme {
+    my (@cmd) = @_;
+    my @lines;
+    $ENV{'LC_ALL'} = 'C';
+    if (open(my $nvme, '-|', 'nvme', @cmd)) {
+        @lines = <$nvme>;
+        close($nvme);
+    } else {
+        # Perl printed a warning about failed exec already.  Ignore
+        # error and return nothing.
+    }
+    @lines;
+}
+
+sub human_to_bytes {
+    my ($str) = @_;
+    my %units = (
+        kB => 1000,
+        MB => 1000_000,
+        GB => 1000_000_000,
+        TB => 1000_000_000_000,
+        PB => 1000_000_000_000_000,    # I wish I had need for this
+    );
+    $str =~ /(\d+(\.\d+)?)\s+(.B)/;
+    int($1 * $units{$3});
+}
+
+sub nvme_list {
+    # Node             SN                   Model                                    Namespace Usage                      Format           FW Rev
+    # ---------------- -------------------- ---------------------------------------- --------- -------------------------- ---------------- --------
+    # /dev/nvme1n1     S464NB0K601188N      Samsung SSD 970 EVO 2TB                  1         695.50  GB /   2.00  TB    512   B +  0 B   1B2QEXE7
+    my %devices;
+    for (run_nvme('list')) {
+        if (m:^(/\S+)\s+(\S+)\s+(\S.*\S)\s{3,}(\d+)\s+(\S+\s+.B)\s+/\s+(\S+\s+.B):) {
+            $devices{$2} = {
+                device    => $1,
+                sn        => $2,
+                model     => $3,
+                namespace => $4,
+                usage     => human_to_bytes($5),
+                capacity  => human_to_bytes($6),
+            };
+        }
+    }
+    \%devices;
+}
+
+sub smart_log {
+    my ($dev) = @_;
+    my %info;
+    for (run_nvme('smart-log', $dev)) {
+        next if /^Smart Log/;
+        if (/(.*?)\s+:\s+(.*)/) {
+            my ($var, $value) = ($1, $2);
+            $var =~ s/\s/_/g;
+            if ($value =~ /^\d+(,\d\d\d)+$/) {
+                $value =~ s/,//g;
+            }
+            $info{lc $var} = $value;
+        }
+    }
+    return \%info;
+}
+
+use Data::Dumper;
+
+my $mode = ($ARGV[0] or "print");
+
+my $list = nvme_list();
+if ($mode eq 'autoconf') {
```

Comment:
Here you need to check for the existence of the nvme executable, or?

Reply:
well ..  I think the current failure mode is OK?

```
  # munin-run nvme autoconf
  Can't exec "nvme": No such file or directory at /etc/munin/plugins/disk/nvme line 89.
  no (no devices to monitor)
```
checking lsmod output is a possibility to give a better suggestion to the admin, but that would tie the plugin even closer to Linux.  is autoconf supposed to be silent on stderr?

Reply:
Yes, it's supposed to be silent on stderr.